mlir/lib/QEC/Transforms/qec_lowering.cpp

```
@@ -0,0 +1,69 @@
+// Copyright 2025 Xanadu Quantum Technologies Inc.
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+
+//     http://www.apache.org/licenses/LICENSE-2.0
+
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#define DEBUG_TYPE "convert-to-qec"
+
+#include "mlir/IR/PatternMatch.h"
+#include "mlir/Pass/Pass.h"
+
+#include "QEC/IR/QECDialect.h"
+#include "QEC/Transforms/Passes.h"
+#include "QEC/Transforms/Patterns.h"
+#include "Quantum/IR/QuantumOps.h"
+
+#include "mlir/Dialect/Affine/IR/AffineOps.h"
+#include "mlir/Dialect/Arith/IR/Arith.h"
+#include "mlir/Dialect/Func/IR/FuncOps.h"
+
+using namespace mlir;
+using namespace catalyst::quantum;
+using namespace catalyst::qec;
+
+namespace catalyst {
+namespace qec {
+
+#define GEN_PASS_DEF_LOWERINGTOQECPASS
+#define GEN_PASS_DECL_LOWERINGTOQECPASS
+#include "QEC/Transforms/Passes.h.inc"
+
+struct LoweringToQECPass : impl::LoweringToQECPassBase<LoweringToQECPass> {
+    using LoweringToQECPassBase::LoweringToQECPassBase;
+
+    void runOnOperation() final
+    {
+        ConversionTarget target(getContext());
+
+        target.addLegalDialect<arith::ArithDialect, func::FuncDialect, qec::QECDialect>();
```

For a partial conversion there is actually no need to add legal dialects, since dialects with "unknown" legality are already allowed through.

I'm actually wondering whether we want this pass to be a partial conversion or a full one. I guess we might want to leave arbitrary classical dialects untouched, in which case a partial conversion will be easier than marking all the known classical dialects legal. However, I imagine that if other ops from the quantum dialect remain in the IR we would want to raise an error 🤔 In that case, we could mark the entire quantum dialect illegal, and selectively add specific legal instructions that are still allowed, like `quantum.init`, `quantum.alloc`, and `quantum.extract`.

---

Yes I think using partial conversion is better. I think we can specify `CustomOp` and `MeasureOp` as illegal. Then,  rest of quantum dialect will be marked as unknown ops, so we can dynamically mark if those are not `CustomOp` and `MeasureOp` will be all legal.